proposals/0383-allow-let-property-wrapper.md

```diff
@@ -0,0 +1,179 @@
+struct Wrapper {
+  var wrappedValue: Int { 0 }
+  init(wrappedValue: Int) {}
+}
+
+struct S {
+  @Wrapper let value: Int // Error: Property wrapper can only be applied to a ‘var’
+}
+```
+Permitting wrapped properties to mimic the rules for other type instances that can be written with either a `var` or a `let` will simplify the Swift language.
+
+Additionally, `let` wrapped properties add code safety where a user wants to expressly remove access to a property’s mutators after initializing the property once or simply does not need a mutable property wrapper. This could be useful for property wrappers that do not change or are reference types.
+
+## Proposed solution
+
+We propose to allow the application of property wrappers to let declared properties, which will permit the wrapper type to be initialized only once without affecting the implementation of the backing storage.
+
+For example, [TSPL](https://docs.swift.org/swift-book/LanguageGuide/Properties.html#ID617) defines a `SmallNumber` property wrapper and applies it to `UnitRectangle` properties:
+```
+@propertyWrapper
+struct SmallNumber {
+  private var maximum: Int
+  private var number: Int
+
+  var wrappedValue: Int {
+    get { return number }
+    set { number = min(newValue, maximum) }
+  }
+
+  init(wrappedValue: Int) {
+    maximum = 12
+    number = min(wrappedValue, maximum)
+  }
+}
+
+struct UnitRectangle {
+  @SmallNumber var height: Int = 1
+  @SmallNumber var width: Int = 1
+}
+```
+Initial values for `height` and `width` are set using the `init(wrappedValue:)` initializer of the `SmallNumber` property wrapper type. To ensure that `height` and `width` are not changed again, we could add logic to the `wrappedValue` setter to check if the property was already initialized and prevent re-assignment. However, this is an inconvenient solution.
+
+Instead, we could declare these properties with a `let`, synthesize a local `let` constant for the backing storage property (prefixed with an underscore), and only allow the property wrapper to be initialized once, passing the assigned value to `init(wrappedValue:)`. Now, rewriting `UnitRectangle`’s properties as `let` constants will translate to:
+```
+private let _height: SmallNumber = SmallNumber(wrappedValue: 1)
+var height: Int {
+  get { return _height.wrappedValue }
+}
+
+private let _weight: SmallNumber = SmallNumber(wrappedValue: 1)
+var weight: Int {
+  get { return _weight.wrappedValue }
+}
+```
+and results in code that is easy to write and understand:
+```
+struct UnitRectangle {
+  @SmallNumber let height: Int = 1
+  @SmallNumber let width: Int = 1
+}
+```
+
+Property wrappers with `let` declarations will be allowed both as members and local declarations, as envisioned by SE-0258 for `var` declared property wrappers. All other property wrapper traits also remain unchanged from SE-0258.
+
+## Detailed design
+
+Here are three examples of how a `let` wrapped property can make current iterations more effortless.
```

Comment:
The examples are part "Motivation" and part "Proposed solution," but I don't think they fit with what's expected in this section.

Consider looking at what SE-0258 puts into this section for some ideas; it doesn't have to be so long of course. My suggestion would be that it bears formally spelling out that, just like before, property wrappers still work by transforming the source code, then showing worked examples of transforms using the surface language. Start with the basics that you demonstrate in the "Proposed solution" section, but work in the questions you've encountered on the pitch thread concerning how the transform will work in the case of `projectedValue`, `init()`, etc. Such questions dominated the existing discussion and it would be helpful that any document presented for review doesn't leave gaps in that understanding, as the review period will get more feedback evaluating the proposed design if it's not spent mostly developing readers' understanding of what that is.

In general, consider incorporating the questions on the pitch thread that are about "_what would happen if..._" in terms of the observed behavior if one tries to set some value directly or indirectly, or "_is it possible to..._" (for instance, the multiphase initialization example) as part of the "Proposed solution" section, and questions about _the mechanics of how_ some code is transformed as part of the "Detailed design."